scan/base.go

```diff
@@ -40,6 +44,378 @@ type base struct {
+	}
+
+	return
+}
+
+//WpCveInfos is for wpvulndb's json
+type WpCveInfos struct {
+	ReleaseDate     string      `json:"release_date"`
+	ChangelogURL    string      `json:"changelog_url"`
+	Status          string      `json:"status"`
+	LatestVersion   string      `json:"latest_version"`
+	LastUpdated     string      `json:"last_updated"`
+	Popular         bool        `json:"popular"`
+	Vulnerabilities []WpCveInfo `json:"vulnerabilities"`
+	Error           string      `json:"error"`
+}
+
+//WpCveInfo is for wpvulndb's json
+type WpCveInfo struct {
+	ID            int        `json:"id"`
+	Title         string     `json:"title"`
+	CreatedAt     string     `json:"created_at"`
+	UpdatedAt     string     `json:"updated_at"`
+	PublishedDate string     `json:"published_date"`
+	VulnType      string     `json:"vuln_type"`
+	References    References `json:"references"`
+	FixedIn       string     `json:"fixed_in"`
+}
+
+//References is for wpvulndb's json
+type References struct {
+	URL     []string `json:"url"`
+	Cve     []string `json:"cve"`
+	Secunia []string `json:"secunia"`
+}
+
+func detectWp(c *base) (vinfos []models.VulnInfo, err error) {
+	var coreVulns []models.VulnInfo
+	if coreVulns, err = detectWpCore(c); err != nil {
+		return
+	}
+	vinfos = append(vinfos, coreVulns...)
+
+	var themeVulns []models.VulnInfo
+	if themeVulns, err = detectWpTheme(c); err != nil {
+		return
+	}
+	vinfos = append(vinfos, themeVulns...)
+
+	var pluginVulns []models.VulnInfo
+	if pluginVulns, err = detectWpPlugin(c); err != nil {
+		return
+	}
+	vinfos = append(vinfos, pluginVulns...)
+
+	return
+}
+
+func detectWpCore(c *base) (vinfos []models.VulnInfo, err error) {
+	cmd := fmt.Sprintf("wp core version --path=%s", c.ServerInfo.WpPath)
+
+	var coreVersion string
+	var r execResult
+	if r = exec(c.ServerInfo, cmd, noSudo); r.isSuccess() {
+		tmp := strings.Split(r.Stdout, ".")
+		coreVersion = strings.Join(tmp, "")
+		coreVersion = strings.TrimRight(coreVersion, "\r\n")
+		if len(coreVersion) == 0 {
+			return
+		}
+	}
+	if !r.isSuccess() {
+		return vinfos, fmt.Errorf("%s", cmd)
+	}
+
+	url := fmt.Sprintf("https://wpvulndb.com/api/v3/wordpresses/%s", coreVersion)
+	token := fmt.Sprintf("Token token=%s", c.ServerInfo.WpToken)
+	var req *http.Request
+	req, err = http.NewRequest("GET", url, nil)
+	if err != nil {
+		return
+	}
+	req.Header.Set("Authorization", token)
+	client := new(http.Client)
+	var resp *http.Response
+	resp, err = client.Do(req)
+	if err != nil {
+		return
+	}
+	body, _ := ioutil.ReadAll(resp.Body)
+	defer resp.Body.Close()
+	if resp.StatusCode != 200 && resp.StatusCode != 404 {
+		return vinfos, fmt.Errorf("status: %s", resp.Status)
+	} else if resp.StatusCode == 404 {
+		var jsonError WpCveInfos
+		if err = json.Unmarshal([]byte(string(body)), &jsonError); err != nil {
+			return
+		}
+		if jsonError.Error == "HTTP Token: Access denied.\n" {
+			return vinfos, fmt.Errorf("wordpress: HTTP Token: Access denied.")
```

Comment:
__[Golint]__ error strings should not be capitalized or end with punctuation or a newline

<sub>posted by Sider</sub>